src/pdf_parser/pdf_utils/postprocess_layout.py

```diff
@@ -0,0 +1,305 @@
+    page_height: int,
+    height_threshold: int = 50,
+    start_pixel_buffer: int = 10,
+    end_pixel_buffer: int = 10,
+) -> List[TextBlock]:
+    """Infer the missing blocks in columns by checking for sufficiently large gaps.
+
+    Args:
+        column_blocks: The text blocks to infer the missing blocks from.
+        page_height: The height of the page in num pixels.
+        height_threshold: The number of pixels for a missing blocks to be inferred (avoiding normal whitespace).
+        start_pixel_buffer: num pixels to ignore at the start of the column (avoiding non-columnar titles etc).
+        end_pixel_buffer: The number of pixels to ignore at the end of the column (heuristic to exclude footers, etc).
+
+    Returns:
+        The text blocks with the inferred missing blocks.
+    """
+    # Make sure the blocks are sorted by y_1.
+    column_blocks = column_blocks.sort(key=lambda b: b.coordinates[1])
+    x1 = min([b.coordinates[0] for b in column_blocks])
+    x2 = max([b.coordinates[2] for b in column_blocks])
+    # Iteratively fill in gaps between subsequent blocks.
+    new_blocks = []
+
+    for ix in range(len(column_blocks)):
+        if ix < len(column_blocks) - 1:
+            y1_new = column_blocks[ix].coordinates[3]
+            y2_new = column_blocks[ix + 1].coordinates[1]
+            height_new = y2_new - y1_new
+            if height_new >= height_threshold:
+                new_block_shape = Rectangle(x1, y1_new, x2, y2_new)
+                new_block = TextBlock(
+                    new_block_shape, type="Inferred from gaps", score=1.0
+                )
+                new_blocks.append(new_block)
+        # If there is a gap at the start of the column starting from the pixel buffer, add a block.
+        if (ix == 0) & (
+            column_blocks[ix].coordinates[1] - start_pixel_buffer > height_threshold
+        ):
+            new_blocks.append(
+                TextBlock(
+                    Rectangle(
+                        x1,
+                        start_pixel_buffer,
+                        x2,
+                        column_blocks[ix].coordinates[1],
+                    ),
+                    type="Inferred from gaps",
+                    score=1.0,
+                )
+            )
+        # If there is a gap at the end of the column ending at the pixel buffer, add a block. For this, we need
+        # the page length.
+        elif (ix == len(column_blocks) - 1) & (
+            page_height - column_blocks[ix].coordinates[3] - end_pixel_buffer
+            > height_threshold
+        ):
+            new_blocks.append(
+                TextBlock(
+                    Rectangle(
+                        x1,
+                        column_blocks[ix].coordinates[3],
+                        x2,
+                        page_height - end_pixel_buffer,
+                    ),
+                    type="Inferred from gaps",
+                    score=1.0,
+                )
+            )
```

Comment:
This is the only post-processing function that has been meaningfully changed. See the conditionals for adding blocks at the start and end of a page (we only dealt with intermediates before).